File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

#define min(a,b) ((a) >(b) ? (b):(a))
#define pr pair<string,int>
using namespace std;

int compare(string begin, string word)
{
    int size = min(begin.size(), word.size());
    int value = 0;
    for(int i=0;i<size;i++)
    {
	    if(begin[i]-word[i] !=0)
	    	value++;
    }

    if(value > 1) return 0;
    else     	  return 1;
}

int solution(string begin, string target, vector<string> words) {
    int answer =0;
    queue<pr> q;
    int level = 0;
    q.push({begin,level});
     
    while(q.size()>0)
    {
        string front = q.front().first;
        int    level = q.front().second;
        if(front == target)
            return level;
        q.pop();
     
        vector<string> word = words;
        for(int i=0; i<word.size();i++)
        {
	        if(compare(front, word[i]) == 1)
            {
                int idx=0;
   	    	    q.push({word[i],level+1});

                while(idx<words.size() && words[idx] != word[i]) idx++;
                words.erase(words.begin()+idx); // 방문한 노드 제거
            }
        }
    }
    return answer;
}

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.java
import java.io.*;
import java.util.*;

class Solution {
    public boolean OneCharDiff(String source, String target){
    int length = source.length();
    int cnt=0;
    for(int i=0;i<length;i++){
        if(source.charAt(i) == target.charAt(i)){
            cnt++;
        }
    }
    
    return (length-cnt==1);
    }
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        Map<String, Boolean> visited = new HashMap<>();
        for(String word:words) visited.put(word,false);
        Queue<Pair<String,Integer>> que = new LinkedList<>();
        que.add(new Pair<>(begin,0));
        while(!que.isEmpty()){
            Pair<String,Integer> pair = que.poll();        
            visited.put(pair.left,true);
            if(pair.left.equals(target)){ answer = pair.right; break; }
            for(String word:words){
                if(!visited.get(word) &&OneCharDiff(pair.left,word)){
                    que.add(new Pair<>(word,pair.right+1));
                    visited.put(word,true);  
                }
            }
        }
        
        return answer;
    }
    
}

class Pair<L,R>{
    L left; R right;
    public Pair(L x, R y){
        this.left = x;
        this.right = y;
    }
}

